### PR TITLE
Adds file locking documentation

### PR DESCRIPTION
**Pull request by derobins**
This will eventually be rolled into the UG what that lands in develop

### CODE REVIEW

File: doc/file-locking.md [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+If you want to check that file locking is on, you'll need to check the fapl
+setting AND check the environment variable, which can override the fapl.
+
+**!!! WARNING !!!**
+
+Disabling the file locks is at your own risk. If more than one writer process
+modifies an HDF5 file at the same time, the file could be corrupted. If a
+reader process reads a file that is being modified by a writer, the writer
+process might attempt to read garbage and encounter errors or even crash.
+
```
Comment (lrknox) [inline doc/file-locking.md:253]: It seems the condition above could satisfy those in lines 256 and 257.  What's the difference?

File: doc/file-locking.md [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+when code needs to perform SWMR-forbidden operations like creating groups
+and datasets before appending data to datasets using SWMR.
+
+Most of the work of this API call involves flushing out the library caches
+in preparation for SWMR access, but there are a few locking operations that
+take place under the hood:
+
+- The file's superblock is marked as in the SWMR writer case, above.
+- For a brief period of time in the call, we convert the exclusive lock to
+  a shared lock. It's unclear why this was done and we'll look into removing
```
Comment (hyoklee) [inline doc/file-locking.md:171]: Shouldn't we remove code that we don't understand? This is a security hole.

File: doc/file-locking.md [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+
+- CMake
+
+    - set `IGNORE_DISABLED_FILE_LOCK` to `ON` to ignore file locks when they
+      are disabled on the file system (default: `ON`).
+    - set `HDF5_USE_FILE_LOCKING` to `OFF` to disable file locks (default: `ON`)
+
+### `H5Pset_file_locking()`
+
+This API call can be used to override the configure defaults. It takes
```
Comment (hyoklee) [inline doc/file-locking.md:206]: How can an API override the configuration defaults? I think it's a bad practice.

File: doc/file-locking.md [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+In the case of:
+
+- A single process accessing a file with write access
+- Any number of processes accessing a file read-only
+
+You can safely disable the file locking scheme.
+
+If you are trying to set up SWMR without the benefit of the file locks, you'll
+just need to be extra careful that you hold to rules for SWMR access.
```
Comment (hyoklee) [inline doc/file-locking.md:262]: Please explain what users should be extra caeful about.

File: doc/file-locking.md [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+## Feature Matrix
+
+The following table indicates which versions of the library support which file
+lock features. 1.13.0 and 1.13.1 are experimental releases (basically glorified
+release candidates) so they are not included here.
+
+**Locks**
+
+- P = POSIX locks only, Windows was a no-op that always succeeded
+- WP = POSIX and Windows locks
```
Comment (hyoklee) [inline doc/file-locking.md:273]: Rewrite "Windows and POSIX locks" to match "WP"

File: doc/file-locking.md [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+- (-) = POSIX no-op lock fails
+- (+) = POSIX no-op lock passes
+
+**Configure Option and Environment Variable**
```
Comment (hyoklee) [inline doc/file-locking.md:277]: I think title should be "Confgiure Option, H5P API, and Environment Variable"

File: doc/file-locking.md [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+One thing that should be immediately apparent to anyone familiar with file
+locking, is that all of these schemes have subtly different semantics. We're
+using file locking in a fairly crude manner, though, and lock use has always
+been optional, so we consider this a lower-order concern.
+
+Locks are implemented at the VFD level via `lock` and `unlock` callbacks. The
+VFDs that implement file locks are: core (w/ backing store), direct, log, sec2,
+and stdio (`flock(2)` locks only). The family, multi, and splitter VFDs invoke
+the lock callback of their underlying sub-files. The onion and MPI-IO VFDs do NOT
```
Comment (hyoklee) [inline doc/file-locking.md:353]: `onion` -> `Onion`

File: doc/file-locking.md [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+
+Lock failures are detected by checking to see if `errno` is set to `ENOSYS`.
+This is not particularly sophisticated and was implemented as a way of working
+around disabled locks on popular parallel file systems.
+
+One other thing to note here is that, in all of the locking schemes we use, the
+file locks do not survive process termination, so you don't have to worry
+about files being locked forever if a process exits abnormally. If a writer
+crashed and the library didn't clear the superblock mark, you can remove it with
+the h5clear command-line tool, which is built with the library.
```
Comment (hyoklee) [inline doc/file-locking.md:366]: Add a `h5clear` example with a command line option. E.g.,

`h5clear -s file_lock_issue_file.h5`

**Review by hyoklee — CHANGES_REQUESTED** [review]
Please see my in-line comments.

**Comment by derobins** [comment]
I'll address the comments when I get a little extra time.

**Pull request merged**